Adding support for EC2 security groups

thebusby opened the pull request:
Hi Michael,

First off I just wanted to thank you for your wonderful EC2 library. I've been using it in my own projects and really appreciate the work you've done.

To better support some of my own work, I've appended some additional code to add/modify/delete EC2 security group definitions in clj-aws-ec2.

I attempted to model all the code based on what I found in the library, and I hope you find this contribution beneficial. As I'm not too familiar with github, or the various etiquette around contributions, I hope you forgive any of my mistakes. 

Hope this helps,
    Alan Busby

thebusby committed:
modified:   src/aws/sdk/ec2.clj
Work to add security group support in progress

thebusby committed:
modified:   project.clj
Updated to point to my fork.
Not sure if this is polite or not, but didn't want to create any confusion.

modified:   src/aws/sdk/ec2.clj
Wrapped up support for creating and updating EC2 security groups.

thebusby committed:
modified:   src/aws/sdk/ec2.clj
Added delete-security-group and cleaned up the descriptions of a few new functions.

thebusby committed:
modified:   src/aws/sdk/ec2.clj
Add security group describe and security group ingress revocation

buildhive commented:
[Michael Rowe » clj-aws-ec2 #47](https://buildhive.cloudbees.com/job/mrowe/job/clj-aws-ec2/47/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

thebusby committed:
modified:   src/aws/sdk/ec2.clj
Added support for creating/deleting EC2 key pairs

buildhive commented:
[Michael Rowe » clj-aws-ec2 #48](https://buildhive.cloudbees.com/job/mrowe/job/clj-aws-ec2/48/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

thebusby committed:
modified:   src/aws/sdk/ec2.clj
Updated create-key-pair and delete-key-pair to just take the name of the key and not a map.

buildhive commented:
[Michael Rowe » clj-aws-ec2 #49](https://buildhive.cloudbees.com/job/mrowe/job/clj-aws-ec2/49/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

thebusby committed:
modified:   project.clj
Updated to point to mrowe's github page.

modified:   src/aws/sdk/ec2.clj
Added describe-key-pairs to get list of EC2 key-pairs

buildhive commented:
[Michael Rowe » clj-aws-ec2 #50](https://buildhive.cloudbees.com/job/mrowe/job/clj-aws-ec2/50/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

mrowe commented on project.clj:
```diff
@@ -1,4 +1,4 @@
-(defproject clj-aws-ec2 "0.2.1"
+(defproject thebusby/clj-aws-ec2 "0.2.2"
```
Please don't include changes to release info/repositories in the patch. :-)

thebusby replied:
I'm sorry about that, it was an early mistake.
I thought you'd reject any changes to project.clj, so I haven't corrected it. What would you prefer?

mrowe replied:
It's easier just to merge the whole pull request than try to manually pick and chose from it, so if you can resubmit it without `project.clj` that would be great.

mrowe commented on src/aws/sdk/ec2.clj:
```diff
@@ -501,3 +514,180 @@
+
+  E.g.:
+
+  (authorize-security-group-ingress cred { :group-id \"sg-abcdefgh\"
```
I know it sounds pedantic, but it would be nice to use valid ids in the examples (i.e. only 0-9 and a-f).

mrowe commented on src/aws/sdk/ec2.clj:
```diff
@@ -501,3 +514,180 @@
+
+(defn describe-security-groups
+  "List all the EC2 security group for the supplied credentials, applying the optional filter if supplied.
+   Returns a list of maps, each containing the contents of a security group."
```
Example usage?

mrowe commented on src/aws/sdk/ec2.clj:
```diff
@@ -501,3 +514,180 @@
+   Including the unencrypted PEM encoded RSA private key.
+   
+   E.g.:
+   (create-key-pair cred \"MyNewKeyName\")"
```
Can you add some info about the data structure that is returned?

mrowe commented on src/aws/sdk/ec2.clj:
```diff
@@ -501,3 +514,180 @@
+
+(defn describe-key-pairs
+  "List all the EC2 key pairs for the supplied credentials.
+  Returns a list of maps, each containing the contents of the key pair.
```
Example of returned data structure?

mrowe commented:
Hi Alan,

Awesome, thanks for the patch! It looks good, I've made a few comments in-line mostly regarding documentation. Please consider them tiny nit-picking, it's great that you went to the trouble of including documentation at all, most people don't. :)

Re GitHub "etiquette", you're doing fine, this is exactly how it's meant to work.

I'm glad you've found the library useful. I've been planning to add more of the API, such as support for volumes, so it's great to have some help with the parts that you've done.

Cheers,
Michael Rowe

thebusby committed:
modified:   project.clj
Hopefully removed any changes made by me

modified:   src/aws/sdk/ec2.clj
Updated per Mrowe's input from https://github.com/mrowe/clj-aws-ec2/pull/6#discussion_r3787368

buildhive commented:
[Michael Rowe » clj-aws-ec2 #51](https://buildhive.cloudbees.com/job/mrowe/job/clj-aws-ec2/51/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

thebusby committed:
modified:   src/aws/sdk/ec2.clj
Includes usage example of describe-security-groups

buildhive commented:
[Michael Rowe » clj-aws-ec2 #52](https://buildhive.cloudbees.com/job/mrowe/job/clj-aws-ec2/52/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

thebusby commented:
Hi Michael, 

Re project.clj; 
I didn't know how to remove it from the pull request so I updated it to reflect what is in your repo.

Re group-id;
All ID's updated to use an example I found earlier in your library, and I tweaked the example names to have a consistent format (Ex. "keypair-name").

Re example usage;
Added to describe-security-groups and describe-key-pairs.

Re create-key-pair data structure returned;
Added example in the doc-string.

Re describe-key-pairs data structure returned;
Added example in the doc-string.

Hope this helps,
   Alan Busby

mrowe commented:
Hi Alan,

Sorry this has taken a while, I've been a bit snowed under. I'm just trying to clean up these new functions to make them a bit more consistent with the rest of the library. Hopefully I'll have something soon.

Cheers,
Michael Rowe

thebusby commented:
No hurry. I am rather occupied myself, but wanted to clean things up and
resubmit properly.

I'm filling in clj-aws-r53, a new IAM library, and a imds library on the
side too.

Wanted to discuss the cred usage though, as I'm finding that instances
using roles should not provide creds. I've been modifying clients to take
creds without access/secret key. Thoughts?

Thanks again!

On Thursday, April 25, 2013, Michael Rowe wrote:

> Hi Alan,
> 
> Sorry this has taken a while, I've been a bit snowed under. I'm just
> trying to clean up these new functions to make them a bit more consistent
> with the rest of the library. Hopefully I'll have something soon.
> 
> Cheers,
> Michael Rowe
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/mrowe/clj-aws-ec2/pull/6#issuecomment-16989497
> .
